src/Polysemy/Scoped.hs

```diff
@@ -26,75 +91,34 @@ import Polysemy.Internal.Tactics (Tactical, runTactics)
 --
 -- @scopedInterpreter@ is a regular interpreter that is called with the @resource@ argument produced by @scope@.
+-- This is mostly useful if you want to reuse an interpreter that you cannot easily rewrite (like from another library).
+-- If you have full control over the implementation, 'interpreteScoped' should be preferred.
 --
 -- /Note/: This function will be called for each action in the program, so if the interpreter allocates any resources,
 -- they will be scoped to a single action. Move them to @withResource@ instead.
 runScoped ::
-  ∀ resource effect r .
-  (∀ x . (resource -> Sem r x) -> Sem r x) ->
+  ∀ param resource effect r .
+  (∀ x . param -> (resource -> Sem r x) -> Sem r x) ->
   (resource -> InterpreterFor effect r) ->
-  InterpreterFor (Scoped resource effect) r
+  InterpreterFor (Scoped param resource effect) r
 runScoped withResource scopedInterpreter =
   run
   where
-    run :: InterpreterFor (Scoped resource effect) r
+    run :: InterpreterFor (Scoped param resource effect) r
     run =
-      interpretH' \ (Weaving effect s wv ex ins) -> case effect of
+      interpretWeaving \ (Weaving effect s wv ex ins) -> case effect of
         Run resource act ->
           scopedInterpreter resource (liftSem $ injWeaving $ Weaving act s (raise . run . wv) ex ins)
-        InScope main ->
-          ex <$> withResource \ resource -> run (wv (main resource <$ s))
+        InScope param main ->
+          withResource param \ resource -> ex <$> run (wv (main resource <$ s))
+{-# inline runScoped #-}
 
 -- |Variant of 'runScoped' in which the resource allocator is a plain action.
 runScopedAs ::
-  ∀ resource effect r .
-  Sem r resource ->
+  ∀ param resource effect r .
+  (param -> Sem r resource) ->
   (resource -> InterpreterFor effect r) ->
-  InterpreterFor (Scoped resource effect) r
+  InterpreterFor (Scoped param resource effect) r
 runScopedAs resource =
-  runScoped \ f -> f =<< resource
-
--- |Variant of 'runScoped' that takes a higher-order handler instead of an interpreter.
-interpretScopedH ::
-  ∀ resource effect r .
-  (∀ x . (resource -> Sem r x) -> Sem r x) ->
-  (∀ r0 x . resource -> effect (Sem r0) x -> Tactical effect (Sem r0) r x) ->
-  InterpreterFor (Scoped resource effect) r
-interpretScopedH withResource scopedHandler =
-  run
-  where
-    run :: InterpreterFor (Scoped resource effect) r
-    run =
-      interpretH' \ (Weaving effect s wv ex ins) -> case effect of
-        Run resource act ->
-          ex <$> runTactics s (raise . run . wv) ins (run . wv) (scopedHandler resource act)
-        InScope main ->
-          ex <$> withResource \ resource -> run (wv (main resource <$ s))
-
--- |Variant of 'runScoped' that takes a handler instead of an interpreter.
-interpretScoped ::
-  ∀ resource effect r .
-  (∀ x . (resource -> Sem r x) -> Sem r x) ->
-  (∀ r0 x . resource -> effect (Sem r0) x -> Sem r x) ->
-  InterpreterFor (Scoped resource effect) r
```

Comment:
This old version is effectively equivalent to what would be `interpretScoped_`, right? Perhaps it's worth keeping it under that moniker? But then it would make sense to duplicate all the interpreters for the non-parametrised version :thinking:

Reply:
yeah that's a dilemma. in particular with the other PR there would be a rather large number of possible combinations.

We could decide that only the most likely variants should be special-cased, like the one you suggest here.

There's also the issue of naming with `interpretScopedAs`, which I chose back then to convey that the bracketing function is a plain action, but now it is parameterized by `param`.

Not sure what to do here, but it's always possible to add convenience stuff later.